pep-0458.txt

```diff
@@ -730,99 +735,97 @@ efficiently transfer consistent snapshots from PyPI.
+updates its *snapshot* metadata hash and version number, increments its own
+version number, sets a new expiration time, signs it with the *timestamp* role
+key, and writes it to *timestamp.json*.
+
+The snapshot process MUST generate consistent snapshots sequentially, reading
+the notifications received from the transaction process(es) from a
+concurrency-safe FIFO queue. Fortunately, the operation of signing is fast
+enough that this may be done a thousand or more times per second.
+
+The following caveats exist in regards to project transactions and immediate
+availability of target files:
+
+- PyPI allows adding files to a project release at arbitrary later times.
+- PyPI may want to make files available for download before all files of the
+  project release have been uploaded.
+
+To solve this, PyPI MAY define variable transaction windows that mark a
+transaction as finished even before all target files of a project release have
+been uploaded. A variable transaction window SHOULD be chosen so that a client,
+who downloads target files of a project release, is not left in an inconsistent
+state (with our without TUF metadata).
+Furthermore, the transaction window MAY only apply to the generation of a TUF
+consistent snapshot. That is, uploaded target files MAY become available for
+download immediately after being uploaded and before TUF has generated a
+consistent snapshot. If a user then downloads those target files in that short
+time frame, the client (e.g. pip) SHOULD warn the user about missing TUF
+protection and advice to abort download and try later.
+
+At any rate, PyPI SHOULD use a `transaction log`__ to record project
+transaction processes and the snapshot queue for auditing and to recover from
+errors after a server failure.
 
 __ https://en.wikipedia.org/wiki/Transaction_log
 
 
 Cleaning up old metadata
 ------------------------
 
-Prior versions of snapshot, targets, and timestamp metadata does not need to
-be kept indefinitely.  (Root files must be indefinitely retained.)
-However, a client that performs an update MUST be able
-to retrieve a consistent set of versions of the files on the repository.
-For example, if a client downloads a snapshot file, it should retrieve the versions
-of targets metadata that existed when that snapshot file was created, even if
-the targets metadata has been updated concurrently with the client requests.
-Fortunately, the use of hash / version delegations handle this case automatically
-since clients request targets unambiguously.  Once no client could reasonably
-be requesting outdated targets, snapshot or timestamp files, those files may be
-removed to save space.  Thus, files that were obsoleted some reasonable time
-in the past (such as 1 hour) may be safely discarded.
+To avoid running out of disk space, by constantly producing new consistent
+snapshots, PyPI SHOULD regularly delete old consistent snapshots, i.e. metadata
```

Comment:
```suggestion
To avoid running out of disk space due to the constant production of new consistent
snapshots, PyPI SHOULD regularly delete older ones, i.e. metadata
```